tree_unique_args/src/subst.rs

```diff
@@ -0,0 +1,79 @@
+use crate::ir::{Constructor, ESort, Purpose};
+use strum::IntoEnumIterator;
+
+fn subst_rule_for_ctor(ctor: Constructor) -> String {
+    if ctor == Constructor::Arg {
+        return "(rewrite (SubstExpr (Arg id) v) v :ruleset always-run)".to_string();
+    }
+
+    // e.g. "(Add x y)"
+    let ctor_pattern = ctor.construct(|field| field.name.to_string());
+
+    // e.g. "(Add (SubstExpr x v) (SubstExpr y v))"
+    let substed_ctor = ctor.construct(|field| match field.purpose {
+        Purpose::Static(_) | Purpose::CapturingId | Purpose::CapturedExpr => field.name.to_string(),
+        Purpose::ReferencingId => panic!("arg case already handled"),
+        Purpose::SubExpr | Purpose::SubListExpr => {
+            let var = field.name;
+            let sort = field.sort().name();
+            format!("(Subst{sort} {var} v)")
+        }
+    });
+
+    let sort = ctor.sort().name();
+    let br = "\n         ";
+    format!("(rewrite (Subst{sort} {ctor_pattern} v){br}{substed_ctor}{br}:ruleset always-run)")
+}
+
+pub(crate) fn subst_rules() -> Vec<String> {
+    ESort::iter()
+        .map(|sort| "(function Subst* (* Expr) * :unextractable)".replace("*", sort.name()))
```

Comment:
use `format!`

Reply:
That's what I did at first, but I changed to `.replace` so we don't have to bind `sort.name()`. I think this is still readable, we often write `Subst*` when talking about families of functions anyway

Reply:
Ok